src/main/kotlin/com/terraformation/backend/report/db/ReportStore.kt

```diff
@@ -0,0 +1,182 @@
+package com.terraformation.backend.report.db
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.kotlin.readValue
+import com.terraformation.backend.auth.currentUser
+import com.terraformation.backend.customer.model.requirePermissions
+import com.terraformation.backend.db.ReportAlreadySubmittedException
+import com.terraformation.backend.db.ReportLockedException
+import com.terraformation.backend.db.ReportNotFoundException
+import com.terraformation.backend.db.ReportNotLockedException
+import com.terraformation.backend.db.default_schema.OrganizationId
+import com.terraformation.backend.db.default_schema.ReportId
+import com.terraformation.backend.db.default_schema.ReportStatus
+import com.terraformation.backend.db.default_schema.tables.daos.ReportsDao
+import com.terraformation.backend.db.default_schema.tables.pojos.ReportsRow
+import com.terraformation.backend.db.default_schema.tables.references.REPORTS
+import com.terraformation.backend.log.perClassLogger
+import com.terraformation.backend.report.ReportService
+import com.terraformation.backend.report.model.ReportBodyModel
+import com.terraformation.backend.report.model.ReportMetadata
+import com.terraformation.backend.report.model.ReportModel
+import java.time.Clock
+import java.time.ZonedDateTime
+import javax.inject.Named
+import org.jooq.DSLContext
+import org.jooq.JSONB
+
+@Named
+class ReportStore(
+    private val clock: Clock,
+    private val dslContext: DSLContext,
+    private val objectMapper: ObjectMapper,
+    private val reportsDao: ReportsDao,
+) {
+  private val log = perClassLogger()
+
+  /**
+   * Fetches a report in whatever format it was written to the database.
+   *
+   * You probably want [ReportService.fetchOneById] instead of this.
+   */
+  fun fetchOneById(reportId: ReportId): ReportModel {
+    requirePermissions { readReport(reportId) }
+
+    val row = reportsDao.fetchOneById(reportId) ?: throw ReportNotFoundException(reportId)
+    val body = objectMapper.readValue<ReportBodyModel>(row.body!!.data())
```

Comment:
Correct, this is non-nullable in the database but Kotlin doesn't know that. All the fields in jOOQ generated classes are nullable because you can use them with arbitrary queries that might only be selecting a subset of the columns. This is a longstanding design problem and there have been a few experiments to try to figure out a good solution, but it's surprisingly tough to solve in a way that doesn't break as many things as it fixes.

Reply:
Ah I see, that makes sense 👍